context/ChainsContext/index.tsx

```diff
@@ -0,0 +1,91 @@
+import { ReactNode, createContext, useContext, useEffect, useReducer } from "react";
+import { setChain, setChains, setChainsError } from "./helpers";
+import { getChain, getChainFromRegistry, getChainItemsFromRegistry } from "./service";
+import { storeChain } from "./storage";
+import { Action, ChainsContextType, State } from "./types";
+
+const ChainsContext = createContext<ChainsContextType | undefined>(undefined);
+
+const chainsReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case "setChains": {
+      return { ...state, chains: action.payload };
+    }
+    case "setChain": {
+      storeChain(action.payload);
+      return { ...state, chain: action.payload };
+    }
+    case "setChainsError": {
+      return { ...state, chainsError: action.payload };
+    }
+    default: {
+      throw new Error("Unhandled action type");
+    }
+  }
+};
+
+interface ChainsProviderProps {
+  readonly children: ReactNode;
+}
+
+export const ChainsProvider = ({ children }: ChainsProviderProps) => {
+  const [state, dispatch] = useReducer(chainsReducer, {
+    chain: getChain(),
+    chains: { mainnets: [], testnets: [] },
+  });
+
+  useEffect(() => {
+    (async function getChainsFromGithubRegistry() {
+      try {
+        const newChainItems = await getChainItemsFromRegistry();
+        setChains(dispatch, newChainItems);
+      } catch (error) {
+        if (error instanceof Error) {
+          setChainsError(dispatch, error.message);
+        } else {
+          setChainsError(dispatch, "Failed to get chains from registry");
+        }
+      }
+    })();
+  }, []);
+
+  useEffect(() => {
+    (async function getChainFromRegistryIfEmpty() {
+      if (!state.chain.chainId && state.chains.mainnets.length && state.chains.testnets.length) {
+        const chainName =
+          state.chain.registryName ||
+          location.pathname.split("/")[1] ||
```

Comment:
Isn't there some elegant way to get this from the framework?

Reply:
I think we should leave it as it is for now.

I tried several approaches and they weren't elegant at all since I needed to pass `NextRouter` coming from `useRouter()` around in many functions (which where not React components and therefore could not access the hook).

It even broke the navigation when changing a network and confirming a redirect, because the redirect does a `router.push()`, and the storing of the chain in the url does a `router.replace()`. And because those methods are `async` and the context's reducer and dispatch functions aren't, I cannot make sure one takes place after another. So it fails randomly.

Reply:
I see, then let's keep it as is. Thanks for checking